include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+   *
+   * This class can then be use with CellDataStorage in the following way:
+   * <code>
+   * CellDataStorage<typename Triangulation<dim,dim>::cell_iterator,MyQData> data_storage;
+   * distributed::ContinuousQuadratureDataTransfer<dim,MyQData> data_transfer(FE_Q<dim>(2),QGauss<dim>(3),QGauss<dim>(4));
+   * //...populate data for all active cells in data_storage
+   * //...mark cells for refinement...
+   * data_transfer.prepare_for_coarsening_and_refinement(triangulation,&data_storage);
+   * triangulation.execute_coarsening_and_refinement();
+   * //...initialize quadrature point data on new cells by calling CellDataStorage::reinit()
+   * data_transfer.interpolate();
+   * </code>
+   * This approach can be extended to quadrature point data with Tensors of
+   * arbitrary order, although with a little bit more work in packing and
+   * unpacking of data inside MyQData class.
+   *
+   * @note Currently coarsening is not supported.
+   *
+   * @note During mesh balancing, children of a single cell can be attributed
+   * to different processors. As a corollary, their parent cells can simultaneously
+   * exist on multiple processors.
+   *
+   * @note The functionality provided by this class can alternatively be achieved
+   * using parallel::distributed::SolutionTransfer . However, that would require
+   * the following steps: (i) create an auxiliary DoFHandler with a (discontinuous Galerkin)
+   * FiniteElement which has enough components to represent all data stored at the
+   * quadrature points; (ii) project the data to the FiniteElement space
+   * and thereby store results in global vectors; (iii) use parallel::distributed::SolutionTransfer
+   * to project FE vectors to the new mesh; and (iv) finally project the
+   * data back to the quadrature points on the new mesh via FEValues class.
+   * The ContinuousQuadratureDataTransfer class aims at simplifying the whole
+   * process by only requiring that the quadrature point data class is derived
+   * from the TransferableQuadraturePointData. Everything else will be done automatically.
+   *
+   * @note This class is not well suited for the transfer of discontinuous
+   * quadrature point data. For example for the elasto-plastic constitutive model
+   * on a cell only some quadrature points may be in the plastic state whereas
+   * others remain elastic. In this scenario using this class to transfer
+   * plastic strain at quadrature points does not make much sense as this field
+   * is discontinuous, whereas cell-based L2-projection performed by this class
+   * will smooth this discontinuity across all child elements.
+   *
+   * @author Denis Davydov, Jean-Paul Pelteret, 2016
+   */
+  template <int dim, typename DataType>
+  class ContinuousQuadratureDataTransfer
+  {
+  public:
+    static_assert(std::is_base_of<TransferableQuadraturePointData, DataType>::value,
+                  "User's DataType class should be derived from QPData");
```

Comment:
Ah, neat -- I had not realized/remembered that `static_assert` can be used at class scope :-)